pytest_inmanta/test_parameter/list_parameter.py

```diff
@@ -87,19 +87,28 @@ def validate(self, raw_value: object) -> Sequence[str]:
 
     def resolve(self, config: "Config") -> Sequence[str]:
         option = config.getoption(self.argument, default=self.default)
-        if option is not None and option is not self.default:
-            # A value is set, and it is not the default one
-            if isinstance(option, collections.abc.Sequence):
-                return self.validate(option)
-            else:
-                return self.validate([option])
-
-        env_var = os.getenv(self.environment_variable)
-        if env_var is not None:
-            # A value is set
-            return self.validate(env_var.split(" "))
-
-        if self.default is not None:
-            return self.default
-
-        raise ParameterNotSetException(self)
+        value_set_using: Optional[str] = None
+
+        try:
+            if option is not None and option is not self.default:
+                # A value is set, and it is not the default one
+                value_set_using = ValueSetBy.CLI
+                if isinstance(option, collections.abc.Sequence):
+                    return self.validate(option)
+                else:
+                    return self.validate([option])
+
+            env_var = os.getenv(self.environment_variable)
+            if env_var is not None:
+                # A value is set
+                value_set_using = ValueSetBy.ENV_VARIABLE
+                return self.validate(env_var.split(" "))
+
+            if self.default is not None:
+                value_set_using = ValueSetBy.DEFAULT_VALUE
+                return self.default
+
+            raise ParameterNotSetException(self)
+
+        finally:
+            self._value_set_using = value_set_using
```

Comment:
Why not just assign to the attribute directly?

Reply:
I can't think of a good reason why not to assign directly, will change it